pkg/chain/ethereum/tbtc.go

```diff
@@ -1236,6 +1236,25 @@ func computeMainUtxoHash(mainUtxo *bitcoin.UnspentTransactionOutput) [32]byte {
 	return mainUtxoHash
 }
 
+func (tc *TbtcChain) ComputeMovingFundsCommitmentHash(
+	targetWallets [][20]byte,
+) [32]byte {
+	return computeMovingFundsCommitmentHash(targetWallets)
+}
+
+func computeMovingFundsCommitmentHash(targetWallets [][20]byte) [32]byte {
+	packedWallets := []byte{}
+
+	for _, wallet := range targetWallets {
+		packedWallets = append(packedWallets, wallet[:]...)
+		// Each wallet hash must be padded with 12 zero bytes following the
+		// actual hash.
+		packedWallets = append(packedWallets, make([]byte, 12)...)
```

Comment:
Why do we need that padding? We have `abi.encodePacked` on the contract side.

Reply:
On te contract side we indeed use `abi.encodePacked`.
But when I ran our unit tests for the Solidity `submitMovingFundsCommitment` and added some debug logs I found that the target wallets are padded by `abi.enocdePacked`:
```
target wallets
0x4b440cb29c80c3f256212d8fdd4f2125366f3c91
0x888f01315e0268bfa05d5e522f8d63f6824d9a96
0xb2a89e53a4227dbe530a52a1c419040735fa636c
packed wallets
0x4b440cb29c80c3f256212d8fdd4f2125366f3c91000000000000000000000000888f01315e0268bfa05d5e522f8d63f6824d9a96000000000000000000000000b2a89e53a4227dbe530a52a1c419040735fa636c000000000000000000000000
movingFundsTargetWalletsCommitmentHash
0x8ba62d1d754a3429e2ff1fb4f523b5fad2b605c873a2968bb5985a625eb96202
```

I decided to use the same exact test data for the [unit test](https://github.com/keep-network/keep-core/blob/99aeec51cec3591053f56f25dff079518b3ad1d8/pkg/chain/ethereum/tbtc_test.go#L341) for `ComputeMovingFundsCommitmentHash‎`.
To make the test pass, I needed to adjust the input for `crypto.Keccak256Hash` with padding.

Reply:
Yes, you are right! I checked https://docs.soliditylang.org/en/latest/abi-spec.html#non-standard-packed-mode which says:

> array elements are padded, but still encoded in-place